airflow/metrics/otel_logger.py

```diff
@@ -0,0 +1,289 @@
+
+    def decr(
+        self,
+        stat: str,
+        count: int = 1,
+        rate: float = 1,
+        tags: Attributes = None,
+    ):
+        """
+        Decrement stat by count.
+
+        :param stat: The name of the stat to decrement.
+        :param count: A positive integer to subtract from current value of stat.
+        :param rate: value between 0 and 1 that represents the sampled rate at
+            which the metric is going to be emitted.
+        :param tags: Tags to append to the stat.
+        """
+        if (count < 0) or (rate < 0):
+            raise ValueError("count and rate must both be positive values.")
+        if rate < 1 and random.random() > rate:
+            return
+
+        if self.metrics_validator.test(stat) and name_is_otel_safe(self.prefix, stat):
+            counter = self.metrics_map.get_counter(f"{self.prefix}.{stat}")
+            counter.add(-count, attributes=tags)
+            return counter
+
+    @validate_stat
+    def gauge(
+        self,
+        stat: str,
+        value: int | float,
+        rate: float = 1,
+        delta: bool = False,
+        *,
+        tags: Attributes = None,
+    ) -> None:
+        warnings.warn("OpenTelemetry Gauges are not yet implemented.")
+        return None
+
+    @validate_stat
+    def timing(
+        self,
+        stat: str,
+        dt: DeltaType,
+        *,
+        tags: Attributes = None,
+    ) -> None:
+        warnings.warn("OpenTelemetry Timers are not yet implemented.")
+        return None
+
+    @validate_stat
+    def timer(
+        self,
+        stat: str | None = None,
+        *args,
+        tags: Attributes = None,
+        **kwargs,
+    ) -> TimerProtocol:
+        warnings.warn("OpenTelemetry Timers are not yet implemented.")
+        return Timer()
+
+
+class MetricsMap:
+    """Stores Otel Instruments."""
+
+    def __init__(self, meter):
+        self.meter = meter
+        self.map = {}
+
+    def clear(self) -> None:
+        self.map.clear()
+
+    def _create_counter(self, name):
+        """Creates a new counter or up_down_counter for the provided name."""
+        otel_safe_name = name[:OTEL_NAME_MAX_LENGTH]
+        if name != otel_safe_name:
+            warnings.warn(
+                f"Metric name `{name}` exceeds OpenTelemetry's name length limit of "
+                f"{OTEL_NAME_MAX_LENGTH} characters and will be truncated to `{otel_safe_name}`."
+            )
+
+        if _is_up_down_counter(name):
+            counter = self.meter.create_up_down_counter(name=otel_safe_name)
+        else:
+            counter = self.meter.create_counter(name=otel_safe_name)
+
+        counter_type = str(type(counter)).split(".")[-1][:-2]
+        logging.debug("--> created %s as type: %s", otel_safe_name, counter_type)
+        return counter
+
+    def get_counter(self, name: str, attributes: Attributes = None):
+        """
+        Returns the counter and creates a new one if it does not exist.
```

Comment:
Just a nit-suggestion, feel free to take it or not :) 
```suggestion
        Returns the counter or creates a new one if it does not exist.
```

Reply:
Interesting.  That is pretty awkward ether way, isn't it.   If the counter didn't exist then it creates it and returns it.  So either way it is returning the counter.  A semicolon would maybe make it less ambiguous?

`Returns the counter; creates a new one if it does not exist.` maybe?

Reply:
Yeah, I'd agree with semicolon :+1: